src/lib.rs

```diff
@@ -17,244 +16,180 @@
-                    elem = -elem;
-                }
-                vec![elem]
-            }
+        if all_full_sizes {
+            self.squeeze_native_field_elements(sizes.len())
+        } else {
+            self.squeeze_field_elements_with_sizes::<Self::CF>(sizes)
         }
-    };
-}
-
-impl_absorbable_signed!(i8, u8);
-impl_absorbable_signed!(i16, u16);
-impl_absorbable_signed!(i32, u32);
-impl_absorbable_signed!(i64, u64);
-impl_absorbable_signed!(i128, u128);
-
-macro_rules! impl_absorbable_size {
-    ($t:ident) => {
-        impl<F: PrimeField> Absorbable<F> for $t {
-            fn to_sponge_bytes(&self) -> Vec<u8> {
-                Absorbable::<F>::to_sponge_bytes(&(*self as u64))
-            }
-
-            fn to_sponge_field_elements(&self) -> Vec<F> {
-                (*self as u64).to_sponge_field_elements()
-            }
-        }
-    };
-}
-
-impl_absorbable_size!(usize);
-impl_absorbable_size!(isize);
-
-impl<F: PrimeField> Absorbable<F> for bool {
-    fn to_sponge_bytes(&self) -> Vec<u8> {
-        vec![(*self as u8)]
-    }
-
-    fn to_sponge_field_elements(&self) -> Vec<F> {
-        vec![F::from(*self)]
-    }
-}
-
-impl<F: PrimeField, A: Absorbable<F>> Absorbable<F> for Option<A> {
-    fn to_sponge_bytes(&self) -> Vec<u8> {
-        let mut output = vec![self.is_some() as u8];
-        if let Some(absorbable) = self {
-            output.extend(absorbable.to_sponge_bytes());
-        };
-        output
-    }
-
-    fn to_sponge_field_elements(&self) -> Vec<F> {
-        let mut output = vec![F::from(self.is_some())];
-        if let Some(absorbable) = self {
-            output.extend(absorbable.to_sponge_field_elements());
-        };
-        output
-    }
-}
-
-impl<F: PrimeField, A: Absorbable<F>> Absorbable<F> for &A {
-    fn to_sponge_bytes(&self) -> Vec<u8> {
-        (*self).to_sponge_bytes()
-    }
-
-    fn to_sponge_field_elements(&self) -> Vec<F> {
-        (*self).to_sponge_field_elements()
     }
 }
 
-/// Individually absorbs each element in a comma-separated list of absorbables into a sponge.
-/// Format is `absorb!(s, a_0, a_1, ..., a_n)`, where `s` is a mutable reference to a sponge
-/// and each `a_i` implements `Absorbable`.
-#[macro_export]
-macro_rules! absorb {
-    ($sponge:expr, $($absorbable:expr),+ ) => {
-        $(
-            CryptographicSponge::absorb($sponge, &$absorbable);
-        )+
-    };
+/// An extension for the interface of a cryptographic sponge.
+/// In addition to operations defined in `CryptographicSponge`, `SpongeExt` can convert itself to
+/// a state, and instantiate itself from state.
+/// TODO: change the name, or move it to main trait?
+pub trait SpongeExt: CryptographicSponge {
+    /// The full state of the cryptographic sponge.
+    type State: Clone;
+    /// Returns a sponge that uses `state`.
+    fn from_state(state: Self::State) -> Self;
+    /// Consumes `self` and returns the state.
+    fn into_state(self) -> Self::State;
 }
```

Comment:
- All sponges should have state. Shall we put those to main trait?
- If not, at least we need to give it a better name. Any idea?

@ValarDragon

Reply:
I think its moreso that you don't have to be exposed to the State if you just use the standard trait

Reply:
Sure, I will add a comment